packages/dds/tree/src/simple-tree/prepareForInsertion.ts

```diff
@@ -0,0 +1,230 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import type {
+	ExclusiveMapTree,
+	SchemaAndPolicy,
+	IForestSubscription,
+	MapTree,
+	UpPath,
+} from "../core/index.js";
+import {
+	type FlexTreeContext,
+	getSchemaAndPolicy,
+	type FlexTreeHydratedContext,
+} from "../feature-libraries/index.js";
+import type { ImplicitAllowedTypes, ImplicitFieldSchema } from "./schemaTypes.js";
+import { type InsertableContent, mapTreeFromNodeData } from "./toMapTree.js";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+import { EmptyKey } from "../core/index.js";
+import { type Mutable, isReadonlyArray } from "../util/index.js";
+import {
+	getKernel,
+	type TreeNode,
+	tryUnhydratedFlexTreeNode,
+	unhydratedFlexTreeNodeToTreeNode,
+} from "./core/index.js";
+
+/**
+ * Prepare content from a user for insertion into a tree.
+ * @remarks
+ * This validates and converts the input, and if necessary invokes {@link prepareContentForHydration}.
+ */
+export function prepareForInsertion<TIn extends InsertableContent | undefined>(
+	data: TIn,
+	schema: ImplicitFieldSchema,
+	destinationContext: FlexTreeContext,
+): ExclusiveMapTree | undefined extends TIn ? undefined : never {
+	return prepareForInsertionContextless(
+		data,
+		schema,
+		getSchemaAndPolicy(destinationContext),
+		destinationContext.isHydrated() ? destinationContext : undefined,
+	);
+}
+
+/**
+ * {@link prepareForInsertion} but batched for array content.
+ * @privateRemarks
+ * TODO:
+ * Experimentally it was determined that making separate calls to prepareContentForHydration for each array item did not work.
+ * This should be understood and fixed or have the factors that cause it clearly documented.
+ * If fixed, this function should be removed, and arrays can just map over prepareForInsertion.
+ */
+export function prepareArrayForInsertion(
```

Comment:
I hope to eventually deduplicate these 3 separate entry points, but at least there are now only three cases, and they are all in one place not scattered throughout the codebase.